### PR TITLE
speech to text js api

### LINKED ISSUE #1754: Speech recognition to recognize answer
Originally reported on Google Code with ID 852

```
By Max:
I wanted to share with you a quick spark of imagination. The last days i
stumbled over Android's SpeechToText-API and thought that this could be
basis for neat feature of Anki on Android. So instead of simply getting
shown the answer-side of a flash-card one may unlock it by saying the
answer. Anki could then check correctness of the answer etc.
Theres and API overview [1] and a demo [2], Japanese is supported as far as
i'm concerned. The regarding post on the developers blog [3]. I know that a
lot of design decision would have to be considered and that this would most
likely cluster Anki with respect to platforms -- quite undesirable.
```

Reported by `nicolas.raoul` on 2011-10-30 04:11:29

### PR DESCRIPTION
**Pull request by krmanik**
## Purpose / Description
_Describe the problem or feature and motivation_

Option to practice speech for language learning, this feature will be very helpful. Also this feature is not cross platform so it will be better to implement it in JS API, then in JS Addons.

## Fixes
* Fixes #1754

## Approach
_How does this change address the problem?_

The speech to text is asynchronous so a callback method used to get the result from SpeechRecognizer. Hook added to get result and parse the result. Option to set language, start and stop SpeechRecognizer.

- https://developer.android.com/reference/android/speech/RecognizerIntent
- https://developer.android.com/reference/android/speech/SpeechRecognizer#startListening(android.content.Intent)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

Tested on device

</img>

## Learning (optional, can help others)
_Describe the research stage_

_Links to blog posts, patterns, libraries or addons used to solve this problem_

- https://developer.android.com/reference/android/speech/RecognizerIntent
- https://developer.android.com/reference/android/speech/SpeechRecognizer

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt [hunk 1]
```diff
@@ -321,6 +324,28 @@ open class AnkiDroidJsAPI(private val activity: AbstractFlashcardViewer) {
...
+                val callback = object : JavaScriptSTT.SpeechRecognitionCallback {
+                    override fun onResult(result: String) {
+                        activity.runOnUiThread {
+                            val apiResult = ApiResult(true, result)
+                            val jsonEncodedString = JSONObject.quote(apiResult.toString())
+                            activity.webView!!.evaluateJavascript("ankiSttResult($jsonEncodedString)", null)
+                        }
+                    }
+                    override fun onError(errorMessage: String) {
+                        activity.runOnUiThread {
```
Comment (ShridharGoel) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt:338]: Same

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt [hunk 2]
```diff
@@ -321,6 +324,28 @@ open class AnkiDroidJsAPI(private val activity: AbstractFlashcardViewer) {
                 activity.flipOrAnswerCard(AbstractFlashcardViewer.EASE_4)
                 convertToByteArray(apiContract, true)
             }
+            "sttSetLanguage" -> convertToByteArray(apiContract, mSpeechRecognizer.setLanguage(apiParams))
+            "sttStart" -> {
+                val callback = object : JavaScriptSTT.SpeechRecognitionCallback {
+                    override fun onResult(result: String) {
+                        activity.runOnUiThread {
```
[the remark this answers is not in the document]
Comment (krmanik) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt]: I have implemented it in next, is it correct implementation?

[the remark this answers is not in the document]
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt]: You can use `lifecycleScope` from `activity`.

File: AnkiDroid/src/main/java/com/ichi2/anki/JavaScriptSTT.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                    else -> "Unknown error"
+                }
+                recognitionCallback?.onError(errorMessage)
+            }
+
+            override fun onResults(results: Bundle?) {
+                val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
+                if (!matches.isNullOrEmpty()) {
+                    val result = matches[0]
+                    recognitionCallback?.onResult(result)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/JavaScriptSTT.kt:110]: Why is this only returning `[0]`?

Comment (krmanik) [reply AnkiDroid/src/main/java/com/ichi2/anki/JavaScriptSTT.kt]: I have updated it to return multiple result as json array.

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (krmanik) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt]: Done

**Review by david-allison — APPROVED** [review]
This looks good!

I'm not at all trusting of the Android audio APIs, but this looks good as a first pass

**Review by ShridharGoel — APPROVED** [review]

**Pull request merged**